File: hydra-web/src/api/mcp.ts
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query';
import { apiClient } from '@/lib/api-client';
import { queryKeys } from '@/lib/query-client';

export type MCPServerCategory =
  | 'infrastructure'
  | 'monitoring'
  | 'version-control'
  | 'databases'
  | 'cloud'
  | 'development'
  | 'other';

export type MCPAuthType = 'none' | 'api_key' | 'bearer';
export type MCPServerStatus = 'unknown' | 'healthy' | 'unhealthy';

export interface MCPServerResponse {
  serverId: string;
  name: string;
  endpoint: string;
  description?: string;
  category: MCPServerCategory;
  authType: MCPAuthType;
  authConfigured: boolean;
  enabled: boolean;
  status: MCPServerStatus;
  lastHealthCheck?: string | null;
  docsUrl?: string | null;
  ownerId: string;
  createdAt: string;
  updatedAt: string;
}

export interface MCPServerListResponse {
  servers: MCPServerResponse[];
  total: number;
}

export interface MCPServerCreate {
  name: string;
  endpoint: string;
  description?: string;
  category?: MCPServerCategory;
  authType?: MCPAuthType;
  authValue?: string;
  enabled?: boolean;
  docsUrl?: string;
}

export interface MCPServerUpdate {
  name?: string;
  endpoint?: string;
  description?: string;
  category?: MCPServerCategory;
  authType?: MCPAuthType;
  authValue?: string;
  enabled?: boolean;
  docsUrl?: string;
}

export interface MCPHealthResponse {
  serverId: string;
  status: MCPServerStatus;
  message: string;
  checkedAt: string;
  tools?: string[];
  resources?: string[];
}

export interface MCPToolInfo {
  name: string;
  description?: string | null;
}

export interface MCPToolsResponse {
  serverId: string;
  tools: MCPToolInfo[];
}

export interface MCPResourceInfo {
  uri: string;
  name?: string | null;
  description?: string | null;
  mimeType?: string | null;
}

export interface MCPResourcesResponse {
  serverId: string;
  resources: MCPResourceInfo[];
}

export interface MCPPromptArgument {
  name: string;
  description?: string | null;
  required: boolean;
}

export interface MCPPromptInfo {
  name: string;
  description?: string | null;
  arguments: MCPPromptArgument[];
}

export interface MCPPromptsResponse {
  serverId: string;
  prompts: MCPPromptInfo[];
}

export interface HydraMCPHealthResponse {
  status: MCPServerStatus;
  message: string;
  checkedAt: string;
  serverName?: string | null;
  version?: string | null;
  toolsCount: number;
  promptsCount: number;
  resourcesCount: number;
  endpoint?: string | null;
}

export function useMCPServers(params?: { category?: MCPServerCategory; enabled?: boolean }) {
  return useQuery({
    queryKey: queryKeys.mcp.servers(),
    queryFn: async () => {
      const response = await apiClient.get<MCPServerListResponse>('/mcp/servers', {
        params: {
          category: params?.category,
          enabled: params?.enabled,
        },
      });
      return response.data;
    },
  });
}

export function useMCPServer(serverId: string) {
  return useQuery({
    queryKey: queryKeys.mcp.serverStatus(serverId),
    queryFn: async () => {
      const response = await apiClient.get<MCPServerResponse>(`/mcp/servers/${serverId}`);
      return response.data;
    },
    enabled: !!serverId,
  });
}

export function useCreateMCPServer() {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: async (data: MCPServerCreate) => {
      const response = await apiClient.post<MCPServerResponse>('/mcp/servers', data);
      return response.data;
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: queryKeys.mcp.servers() });
    },
  });
}

export function useUpdateMCPServer() {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: async ({ serverId, data }: { serverId: string; data: MCPServerUpdate }) => {
      const response = await apiClient.put<MCPServerResponse>(
        `/mcp/servers/${serverId}`,
        data
      );
      return response.data;
    },
    onSuccess: (_data, { serverId }) => {
      queryClient.invalidateQueries({ queryKey: queryKeys.mcp.serverStatus(serverId) });
      queryClient.invalidateQueries({ queryKey: queryKeys.mcp.servers() });
    },
  });
}

export function useDeleteMCPServer() {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: async (serverId: string) => {
      const response = await apiClient.delete(`/mcp/servers/${serverId}`);
      return response.data;
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: queryKeys.mcp.servers() });
    },
  });
}

export function useMCPServerHealth(serverId: string) {
  return useQuery({
    queryKey: [...queryKeys.mcp.serverStatus(serverId), 'health'],
    queryFn: async () => {
      const response = await apiClient.get<MCPHealthResponse>(
        `/mcp/servers/${serverId}/health`
      );
      return response.data;
    },
    enabled: !!serverId,
    refetchInterval: 30000,
  });
}

export function useCheckMCPServerHealth() {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: async (serverId: string) => {
      const response = await apiClient.get<MCPHealthResponse>(
        `/mcp/servers/${serverId}/health`
      );
      return response.data;
    },
    onSuccess: (_data, serverId) => {
      queryClient.invalidateQueries({ queryKey: queryKeys.mcp.serverStatus(serverId) });
    },
  });
}

export function useMCPServerTools(serverId: string) {
  return useQuery({
    queryKey: queryKeys.mcp.tools(serverId),
    queryFn: async () => {
      const response = await apiClient.get<MCPToolsResponse>(
        `/mcp/servers/${serverId}/tools`
      );
      return response.data;
    },
    enabled: !!serverId,
  });
}

export function useMCPServerResources(serverId: string) {
  return useQuery({
    queryKey: queryKeys.mcp.resources(serverId),
    queryFn: async () => {
      const response = await apiClient.get<MCPResourcesResponse>(
        `/mcp/servers/${serverId}/resources`
      );
      return response.data;
    },
    enabled: !!serverId,
  });
}

export function useMCPServerPrompts(serverId: string) {
  return useQuery({
    queryKey: queryKeys.mcp.prompts(serverId),
    queryFn: async () => {
      const response = await apiClient.get<MCPPromptsResponse>(
        `/mcp/servers/${serverId}/prompts`
      );
      return response.data;
    },
    enabled: !!serverId,
  });
}

// ============================================================================
// Dedicated Hydra MCP Hooks
// ============================================================================

export function useHydraMCPHealth() {
  return useQuery({
    queryKey: queryKeys.mcp.hydra.health(),
    queryFn: async () => {
      const response = await apiClient.get<HydraMCPHealthResponse>('/mcp/hydra/health');
      return response.data;
    },
    refetchInterval: 30000, // Poll every 30 seconds
  });
}

export function useHydraMCPTools() {
  return useQuery({
    queryKey: queryKeys.mcp.hydra.tools(),
    queryFn: async () => {
      const response = await apiClient.get<MCPToolsResponse>('/mcp/hydra/tools');
      return response.data;
    },
  });
}

export function useHydraMCPPrompts() {
  return useQuery({
    queryKey: queryKeys.mcp.hydra.prompts(),
    queryFn: async () => {
      const response = await apiClient.get<MCPPromptsResponse>('/mcp/hydra/prompts');
      return response.data;
    },
  });
}

File: hydra-api/hydra/api/v1/models/mcp.py
"""MCP server configuration models."""

from datetime import datetime
from enum import StrEnum

from pydantic import BaseModel, ConfigDict, Field


class MCPServerCategory(StrEnum):
    """MCP server category."""

    INFRASTRUCTURE = "infrastructure"
    MONITORING = "monitoring"
    VERSION_CONTROL = "version-control"
    DATABASES = "databases"
    CLOUD = "cloud"
    DEVELOPMENT = "development"
    OTHER = "other"


class MCPAuthType(StrEnum):
    """MCP server authentication type."""

    NONE = "none"
    API_KEY = "api_key"
    BEARER = "bearer"


class MCPServerStatus(StrEnum):
    """MCP server connection status."""

    UNKNOWN = "unknown"
    HEALTHY = "healthy"
    UNHEALTHY = "unhealthy"


class MCPServerCreate(BaseModel):
    """Request to create an MCP server configuration."""

    name: str = Field(min_length=1, max_length=128)
    endpoint: str = Field(min_length=1, description="MCP server endpoint URL")
    description: str | None = Field(default=None, max_length=1024)
    category: MCPServerCategory = Field(default=MCPServerCategory.OTHER)
    auth_type: MCPAuthType = Field(default=MCPAuthType.NONE, alias="authType")
    auth_value: str | None = Field(
        default=None,
        alias="authValue",
        description="API key or bearer token (encrypted at rest)",
    )
    enabled: bool = Field(default=True)
    docs_url: str | None = Field(default=None, alias="docsUrl")

    model_config = ConfigDict(populate_by_name=True)


class MCPServerUpdate(BaseModel):
    """Request to update an MCP server configuration."""

    name: str | None = Field(default=None, max_length=128)
    endpoint: str | None = Field(default=None)
    description: str | None = Field(default=None, max_length=1024)
    category: MCPServerCategory | None = None
    auth_type: MCPAuthType | None = Field(default=None, alias="authType")
    auth_value: str | None = Field(default=None, alias="authValue")
    enabled: bool | None = None
    docs_url: str | None = Field(default=None, alias="docsUrl")

    model_config = ConfigDict(populate_by_name=True)


class MCPServerResponse(BaseModel):
    """MCP server configuration response."""

    server_id: str = Field(alias="serverId")
    name: str
    endpoint: str
    description: str | None = None
    category: MCPServerCategory
    auth_type: MCPAuthType = Field(alias="authType")
    auth_configured: bool = Field(alias="authConfigured")
    enabled: bool
    status: MCPServerStatus = Field(default=MCPServerStatus.UNKNOWN)
    last_health_check: datetime | None = Field(default=None, alias="lastHealthCheck")
    docs_url: str | None = Field(default=None, alias="docsUrl")
    owner_id: str = Field(alias="ownerId")
    created_at: datetime = Field(alias="createdAt")
    updated_at: datetime = Field(alias="updatedAt")

    model_config = ConfigDict(populate_by_name=True)


class MCPServerListResponse(BaseModel):
    """List of MCP server configurations response."""

    servers: list[MCPServerResponse]
    total: int
    limit: int
    offset: int


class MCPHealthResponse(BaseModel):
    """MCP server health check response."""

    server_id: str = Field(alias="serverId")
    status: MCPServerStatus
    message: str
    checked_at: datetime = Field(alias="checkedAt")
    tools: list[str] | None = Field(default=None, description="Available tools if healthy")
    resources: list[str] | None = Field(default=None, description="Available resources if healthy")

    model_config = ConfigDict(populate_by_name=True)


class MCPToolInfo(BaseModel):
    """Information about an MCP tool."""

    name: str
    description: str | None = None


class MCPToolsResponse(BaseModel):
    """Response for listing MCP server tools."""

    server_id: str = Field(alias="serverId")
    tools: list[MCPToolInfo]

    model_config = ConfigDict(populate_by_name=True)


class MCPResourceInfo(BaseModel):
    """Information about an MCP resource."""

    uri: str
    name: str | None = None
    description: str | None = None
    mime_type: str | None = Field(default=None, alias="mimeType")

    model_config = ConfigDict(populate_by_name=True)


class MCPResourcesResponse(BaseModel):
    """Response for listing MCP server resources."""

    server_id: str = Field(alias="serverId")
    resources: list[MCPResourceInfo]

    model_config = ConfigDict(populate_by_name=True)


class MCPPromptArgument(BaseModel):
    """Argument definition for an MCP prompt."""

    name: str
    description: str | None = None
    required: bool = False


class MCPPromptInfo(BaseModel):
    """Information about an MCP prompt."""

    name: str
    description: str | None = None
    arguments: list[MCPPromptArgument] = Field(default_factory=list)


class MCPPromptsResponse(BaseModel):
    """Response for listing MCP server prompts."""

    server_id: str = Field(alias="serverId")
    prompts: list[MCPPromptInfo]

    model_config = ConfigDict(populate_by_name=True)


class HydraMCPHealthResponse(BaseModel):
    """Dedicated health response for the built-in Hydra MCP server."""

    status: MCPServerStatus
    message: str
    checked_at: datetime = Field(alias="checkedAt")
    server_name: str | None = Field(default=None, alias="serverName")
    version: str | None = None
    tools_count: int = Field(default=0, alias="toolsCount")
    prompts_count: int = Field(default=0, alias="promptsCount")
    resources_count: int = Field(default=0, alias="resourcesCount")
    endpoint: str | None = None

    model_config = ConfigDict(populate_by_name=True)
